### PR TITLE
Quantized Max Pool op

### PR DESCRIPTION
**Pull request by z-a-f**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#20474 Quantized Max Pool op**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D15327923/)

parallel implementaiton of the MaxPool (no ReLU).

Differential Revision: [D15327923](https://our.internmc.facebook.com/intern/diff/D15327923/)

### CODE REVIEW

File: test/test_quantized.py [hunk 1]
```diff
@@ -190,6 +190,31 @@ def test_qadd_relu_different_qparams(self):
         np.testing.assert_equal(qCrelu, qCrelu_hat.int_repr(),
                                 "Quantized addition with ReLU failed.")
 
+    """Tests max pool operation on quantized tensors."""
```
Comment (raghuramank100) [inline test/test_quantized.py]: In addition to these tests, please expose the op as part of nn.quantized.functional by modifying torch/nn/quantized/functional.py.
Also, add corresponding tests for the functional at: caffe2/test/test_nn_quantized.py

File: test/test_quantized.py [hunk 2]
```diff
@@ -190,6 +190,31 @@ def test_qadd_relu_different_qparams(self):
...
+        qa = a.quantize_linear(scale=scale, zero_point=zero_point, dtype=torch.qint8)
+
+        a_hat = qa.dequantize()
+        a_pool = F.max_pool2d(a_hat, kernel_size=k, stride=s, padding=p, dilation=d)
+
+        qa_pool_hat = q_max_pool(qa, kernel_size=k, stride=s, padding=p, dilation=d)
+        a_pool_hat = qa_pool_hat.dequantize()
+
+        np.testing.assert_equal(a_pool.numpy(), a_pool_hat.numpy())
+
```
Comment (raghuramank100) [inline test/test_quantized.py]: Can we ensure that quantized ops have coverage comparable to the floating point ops? Is randomized testing done in this case (@dzhulgakov ?)

Comment (zafartahirov) [reply test/test_quantized.py]: I introduced the randomized tests.

File: aten/src/ATen/native/quantized/cpu/qpool.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+  Tensor qy = at::_empty_affine_quantized(
+    oSizes,
+    qx.options().dtype(toQIntType(qx.scalar_type())),
+    qx.q_scale().toDouble(),
+    qx.q_zero_point().toLong());
+  Q* qxd;
+  if (qx.is_contiguous()) {
+    qxd = qx.data<Q>();
+  } else {
+    qxd = qx.contiguous().data<Q>();
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qpool.cpp]: this is a bug. `.contiguous()` returns a new tensor that immediately gets destroyed but you keep its memory pointer. Also it'd return you the same instance if it's already contigutous.

So just:
```
qxc = qx.contiguous();
Q* qxd = qxc.data<Q>();
```

File: test/test_quantized.py (hunk 1 above)
Comment (zafartahirov) [reply test/test_quantized.py]: Resolving this, as it's gonna be part of a different diff

File: aten/src/ATen/native/quantized/cpu/qpool.cpp [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+namespace at { namespace native {
+namespace {
+
+/* Creates the output shape using the input parameters for the axis.
+
+Args:
+  input_size, kernel_size, padding, stride: Appropriate parameter
+  ceil_mode: use ceiling instead of floor while computing the output shape.
+*/
+inline int64_t pooling_output_shape(int64_t input_size, int64_t kernel_size,
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qpool.cpp]: can you just include the function from `THNN/generic/pooling_shape.h`?

Comment (zafartahirov) [reply aten/src/ATen/native/quantized/cpu/qpool.cpp]: Oh, I didn't know that was an option! :) Will try now!

**Review by raghuramank100 — APPROVED** [review]

**Comment by zafartahirov** [comment]
@raghuramank100 Can you stamp it in the diff view, please?

**Pull request closed without merging**